src/storage/table/row_group.cpp

```diff
@@ -613,9 +636,45 @@ void RowGroup::TemplatedScan(TransactionData transaction, CollectionScanState &s
 						continue;
 					}
 					auto scan_idx = filter.scan_column_index;
-					auto &col_data = GetColumn(filter.table_column_index);
-					col_data.Select(transaction, state.vector_index, state.column_scans[scan_idx],
-					                result.data[scan_idx], sel, approved_tuple_count, filter.filter);
+					auto column_idx = filter.table_column_index;
+
+					if (column_idx == COLUMN_IDENTIFIER_ROW_ID) {
+
+						// If we have a selection vector, we need to generate the row ids for more than "count" rows
+						const auto seq_count = sel.IsSet() ? max_count : count;
+
+						// We do another quick statistics scan for row ids here
+						const auto rowid_start = this->start + current_row;
+						const auto rowid_end = this->start + current_row + seq_count;
+						const auto prune_result = CheckRowIdFilter(filter.filter, rowid_start, rowid_end);
+						if (prune_result == FilterPropagateResult::FILTER_ALWAYS_FALSE) {
+							// We can just break out of the loop here.
+							approved_tuple_count = 0;
+							break;
+						}
+
+						// Generate row ids
+						D_ASSERT(result.data[i].GetType().InternalType() == ROW_TYPE);
+
+						// Create sequence for row ids
+						result.data[i].Sequence(UnsafeNumericCast<int64_t>(this->start + current_row), 1, seq_count);
```

Comment:
Instead of creating a sequence when we have a selection vector, can we create a flat vector instead (unifying the code that does that below, i.e.:)

```cpp
				if (column.IsRowIdColumn()) {
					D_ASSERT(result.data[i].GetType().InternalType() == PhysicalType::INT64);
					result.data[i].SetVectorType(VectorType::FLAT_VECTOR);
					auto result_data = FlatVector::GetData<int64_t>(result.data[i]);
					for (size_t sel_idx = 0; sel_idx < approved_tuple_count; sel_idx++) {
						result_data[sel_idx] =
						    UnsafeNumericCast<int64_t>(this->start + current_row + sel.get_index(sel_idx));
					}

```